src/api/object.rs

```diff
@@ -0,0 +1,132 @@
+use crate::{
+    api::metadata::{ListMeta, Metadata},
+    ErrorResponse,
+};
+use serde::Deserialize;
+use std::fmt::Debug;
+
+/// A raw event returned from a watch query
+///
+/// Note that a watch query returns many of these as newline separated json.
+#[derive(Deserialize, Serialize, Clone)]
+#[serde(tag = "type", content = "object", rename_all = "UPPERCASE")]
+pub enum WatchEvent<K>
+where
+    K: Clone + Metadata,
+{
+    Added(K),
+    Modified(K),
+    Deleted(K),
+    Error(ErrorResponse),
+}
+
+impl<K> Debug for WatchEvent<K>
+where
+    K: Clone + Metadata,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match &self {
+            WatchEvent::Added(_) => write!(f, "Added event"),
+            WatchEvent::Modified(_) => write!(f, "Modified event"),
+            WatchEvent::Deleted(_) => write!(f, "Deleted event"),
+            WatchEvent::Error(e) => write!(f, "Error event: {:?}", e),
+        }
+    }
+}
+
+// -------------------------------------------------------
+
+/// A generic kubernetes object list
+///
+/// This is used instead of a full struct for `DeploymentList`, `PodList`, etc.
+/// Kubernetes' API [always seem to expose list structs in this manner](https://docs.rs/k8s-openapi/0.4.0/k8s_openapi/apimachinery/pkg/apis/meta/v1/struct.ObjectMeta.html?search=List).
+///
+/// Note that this is only used internally within reflectors and informers,
+/// and is generally produced from list/watch/delete collection queries on an `Resource`.
+///
+/// This is now equivalent to k8s_openapi::List<T>, but with extra convenience impls
```

Comment:
Is this machinery worth it to be able to say `for x in list.iter() {}` rather than `for x in list.items.iter() {}`? Personally, I almost prefer the latter..

Reply:
if feels nicer to beginners who can do node like things like:

```rust
for pod in api.list() {
     println!("{:?}", pod)
}
```

and it's common pattern to implement for wrapper types like this. I don't think there's a real cost other than the couple of extra lines in here?

Reply:
For one, `k8s_openapi::List` implements a bunch of traits that `kube::api::List` doesn't (`Debug`, `PartialEq`, `Serialize`, ...).

Reply:
Yeah. that is true. Probably worth moving this over as well.